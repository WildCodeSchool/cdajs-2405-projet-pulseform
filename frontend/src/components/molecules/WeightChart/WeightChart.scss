@use "@assets/variables" as *;

.weight-chart {
  width: 100%;

  &__title {
    @include h2-phone;
    display: flex;
    justify-content: center;
    color: $primary;
    width: 100%;
  }
}
