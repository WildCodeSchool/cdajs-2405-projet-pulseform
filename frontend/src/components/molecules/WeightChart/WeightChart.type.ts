export interface DataPoint {
  month: string;
  poids: number | null;
}
