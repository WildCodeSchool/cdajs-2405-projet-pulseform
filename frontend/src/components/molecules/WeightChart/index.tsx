import { useTranslation } from "react-i18next";
import {
  Line,
  LineChart,
  ReferenceLine,
  ResponsiveContainer,
  Tooltip,
  type TooltipProps,
  XAxis,
} from "recharts";

import { useGetUserByIdWithWeights } from "@hooks/useUsers";
import type { DataPoint } from "./WeightChart.type";
import "./WeightChart.scss";

function WeightChart({ userId }: { userId: number }) {
  const { t } = useTranslation();

  const { loading, error, userWeight } = useGetUserByIdWithWeights(userId);

  // todo nelson a finir
  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error: {error.message}</p>;
  if (!userWeight) return <p>No data available</p>;

  const monthsOrder = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
  ];

  // Mois actuel
  const currentMonth = new Date().toLocaleString("en-US", { month: "short" });

  // Trouver l'index du mois actuel dans le tableau
  const currentMonthIndex = monthsOrder.findIndex(
    (month) => month === currentMonth,
  );

  // Sélection des 6 derniers mois
  const last6Months: string[] = [];
  for (let i = 6; i >= 0; i--) {
    const monthIndex = (currentMonthIndex - i + 12) % 12;
    last6Months.push(monthsOrder[monthIndex]);
  }

  // Construire les données complètes avec des valeurs manquantes si besoin
  const completeData: DataPoint[] = last6Months.map((month) => {
    const entry = userWeight.find((d) => d.month === month);
    return entry || { month, weight: null }; // Ajoute le mois s'il manque
  });

  // Fonction pour traduire les mois
  const getMonthLabel = (month: string): string => {
    return t(month.toUpperCase());
  };

  const CustomTooltip = ({ active, payload }: TooltipProps<number, string>) => {
    if (active && payload && payload.length) {
      return (
        <div
          style={{
            background: "#132945",
            color: "#fff",
            padding: "5px 10px",
            borderRadius: "10px",
            fontSize: "12px",
          }}
        >
          {payload[0].value !== null ? `${payload[0].value} kg` : t("NO_DATA")}
        </div>
      );
    }
    return null;
  };

  return (
    <section className="weight-chart">
      {!loading && !error && userWeight.length > 0 && (
        <div style={{ height: 220, position: "relative" }}>
          <div
            style={{
              display: "flex",
              justifyContent: "center",
              fontSize: "18px",
              fontWeight: "bold",
              padding: "5px 10px",
            }}
          >
            <span className="weight-chart__title">{t("WEIGHT_KG")}</span>
          </div>

          <ResponsiveContainer width="100%" height={180}>
            <LineChart data={completeData}>
              {completeData.map((entry) => (
                <ReferenceLine
                  key={entry.month}
                  x={entry.month}
                  stroke="#D1D5DB"
                  strokeDasharray="3 3"
                />
              ))}

              <XAxis
                dataKey="month"
                tick={({ x, y, payload }) => (
                  <text
                    x={x}
                    y={y + 10}
                    fill="#4B5563"
                    fontSize={12}
                    fontWeight={"bold"}
                    textAnchor="middle"
                    style={{
                      whiteSpace: "nowrap",
                      transform: "translateY(4px)",
                    }}
                  >
                    {getMonthLabel(payload.value)}
                  </text>
                )}
                tickLine={false}
                axisLine={false}
                interval={0}
                padding={{ left: 20, right: 20 }}
              />
              <Tooltip content={<CustomTooltip />} cursor={false} />
              <Line
                type="monotone"
                dataKey="weight"
                stroke="#0F172A"
                strokeWidth={2}
                connectNulls={true}
              />
            </LineChart>
          </ResponsiveContainer>
        </div>
      )}
    </section>
  );
}

export default WeightChart;
