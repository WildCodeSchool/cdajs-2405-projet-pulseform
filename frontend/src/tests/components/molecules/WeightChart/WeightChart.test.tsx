import WeightChart from "@components/molecules/WeightChart";
import { render } from "@testing-library/react";
import { vi } from "vitest";

global.ResizeObserver = class {
  observe() {}
  unobserve() {}
  disconnect() {}
};

describe("WeightChart", () => {
  beforeAll(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date("2025-03-07T12:00:00Z"));
  });

  afterAll(() => {
    vi.useRealTimers();
  });

  it("renders the WeightChart for require props", () => {
    const tree = render(<WeightChart {...requireProps} />);
    expect(tree).toMatchSnapshot();
  });

  it("renders the WeightChart has not data", () => {
    const tree = render(<WeightChart weightChartData={[]} />);
    expect(tree).toMatchSnapshot();
  });

  it("renders the WeightChart has not full data", () => {
    const tree = render(<WeightChart {...dataNotFullProps} />);
    expect(tree).toMatchSnapshot();
  });
});

const dataNotFullProps = {
  weightChartData: [
    { month: "May", poids: 65 },
    { month: "Jun", poids: null },
    { month: "Jul", poids: 64 },
    { month: "Aug", poids: 66 },
    { month: "Sep", poids: null },
    { month: "Oct", poids: 71 },
    { month: "Nov", poids: 70 },
    { month: "Dec", poids: null },
  ],
};

const requireProps = {
  weightChartData: [
    { month: "May", poids: 65 },
    { month: "Jun", poids: 68 },
    { month: "Jul", poids: 64 },
    { month: "Aug", poids: 66 },
    { month: "Sep", poids: 69 },
    { month: "Oct", poids: 71 },
    { month: "Nov", poids: 70 },
    { month: "Dec", poids: 50 },
    { month: "Mar", poids: 115 },
  ],
};
