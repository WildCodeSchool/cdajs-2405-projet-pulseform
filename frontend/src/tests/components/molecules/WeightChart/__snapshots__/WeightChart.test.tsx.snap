// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WeightChart > renders the WeightChart for require props 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="width: 320px; height: 220px; position: relative;"
      >
        <div
          style="display: flex; justify-content: space-between; font-size: 18px; font-weight: bold; padding: 5px 10px;"
        >
          <span>
            WEIGHT
          </span>
        </div>
        <div
          class="recharts-responsive-container"
          style="width: 100%; height: 180px; min-width: 0;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="width: 320px; height: 220px; position: relative;"
    >
      <div
        style="display: flex; justify-content: space-between; font-size: 18px; font-weight: bold; padding: 5px 10px;"
      >
        <span>
          WEIGHT
        </span>
      </div>
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 180px; min-width: 0;"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`WeightChart > renders the WeightChart has not data 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="width: 320px; height: 220px; position: relative;"
      >
        <div
          style="display: flex; justify-content: space-between; font-size: 18px; font-weight: bold; padding: 5px 10px;"
        >
          <span>
            WEIGHT
          </span>
        </div>
        <div
          class="recharts-responsive-container"
          style="width: 100%; height: 180px; min-width: 0;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="width: 320px; height: 220px; position: relative;"
    >
      <div
        style="display: flex; justify-content: space-between; font-size: 18px; font-weight: bold; padding: 5px 10px;"
      >
        <span>
          WEIGHT
        </span>
      </div>
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 180px; min-width: 0;"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`WeightChart > renders the WeightChart has not full data 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="width: 320px; height: 220px; position: relative;"
      >
        <div
          style="display: flex; justify-content: space-between; font-size: 18px; font-weight: bold; padding: 5px 10px;"
        >
          <span>
            WEIGHT
          </span>
        </div>
        <div
          class="recharts-responsive-container"
          style="width: 100%; height: 180px; min-width: 0;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="width: 320px; height: 220px; position: relative;"
    >
      <div
        style="display: flex; justify-content: space-between; font-size: 18px; font-weight: bold; padding: 5px 10px;"
      >
        <span>
          WEIGHT
        </span>
      </div>
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 180px; min-width: 0;"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
